infrastructures/infrastructure-gce/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/GCEInfrastructure.java

```diff
@@ -255,63 +274,171 @@ private GCECredential getCredentialFromJsonKeyFile(byte[] credsFile) {
+                } finally {
+                    dynamicAcquireLock.unlock();
+                }
+            } else {
+                logger.info("Infrastructure is busy, acquireNodes skipped.");
+            }
+        });
+    }
+
+    /**
+     * deploy {@code nbInstancesToDeploy} instances with  {@code numberOfNodesPerInstance} nodes on each instance
+     * @param nbInstancesToDeploy number of instances to deploy
+     */
+    private void deployInstancesWithFullNodes(int nbInstancesToDeploy) {
+        logger.info(String.format("Deploying %d instances with %d nodes on each instance.",
+                                  nbInstancesToDeploy,
+                                  numberOfNodesPerInstance));
+
         connectorIaasController.waitForConnectorIaasToBeUP();
 
-        connectorIaasController.createInfrastructure(getInfrastructureId(),
-                                                     gceCredential.clientEmail,
-                                                     gceCredential.privateKey,
-                                                     null,
-                                                     DESTROY_INSTANCES_ON_SHUTDOWN);
-
-        // the initial scripts to be executed on each node requires the identification of the instance (i.e., instanceTag), which can be retrieved through its hostname on each instance.
-        final String instanceTagOnNode = "$HOSTNAME";
-        final String nodeNameOnNode = "$HOSTNAME";
-        List<String> scripts = linuxInitScriptGenerator.buildScript(instanceTagOnNode,
-                                                                    getRmUrl(),
-                                                                    rmHostname,
-                                                                    INSTANCE_TAG_NODE_PROPERTY,
-                                                                    additionalProperties,
-                                                                    nodeSource.getName(),
-                                                                    nodeNameOnNode,
-                                                                    numberOfNodesPerInstance);
+        createInfrastructureIfNeeded();
+
+        List<String> nodeStartCmds = buildNodeStartScripts(numberOfNodesPerInstance);
+
+        Set<String> instancesIds = createInstanceWithNodesStartCmd(nbInstancesToDeploy, nodeStartCmds);
+
+        declareDeployingNodes(instancesIds, numberOfNodesPerInstance, nodeStartCmds.toString());
+    }
 
+    // Calculate the required number of instances to deploy, with numberOfNodesPerInstance nodes on each instance,
+    // while conforming to the constraint of max instances number and max nodes number
+    // Note we may deploy more nodes then required as long as it not exceeds max nodes number.
+    private int calNumberOfInstancesToDeploy(final int numberOfNodes, Map<String, ?> dynamicPolicyParameters) {
```

Comment:
I think this method could be defined in AbstractAddonInfrastructure.

In order to do so the method should take two more parameters, such as:

private int calculateNumberOfInstancesToDeploy(final int numberOfNodesRequested, Map<String, ?> dynamicPolicyParameters, int numberOfNodesPerInstance, int maximumNumberOfInstances)

It would be also possible to refactor and pull up in AbstractAddonInfrastructure the two common parameters which appear in all node sources (totalNumberOfInstances, numberOfNodesPerInstance) but that's not mandatory

Reply:
Done. I've added two parameters. (I think it would be more clear to keep the parameters in each infrastructure.)